osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -18,146 +23,296 @@
+        return this.doRetrieveSecrets(deviceIdentification, secretTypes, SecretStatus.NEW);
+    }
+
+    private List<TypedSecret> doRetrieveSecrets(final String deviceIdentification, final List<SecretType> secretTypes,
+            SecretStatus status) {
+        return this.retrieveAesSecrets(deviceIdentification, secretTypes, status).stream().map(this::reencryptAes2Rsa)
+                   .map(EncryptedTypedSecret::toTypedSecret).collect(Collectors.toList());
+    }
+
+    private List<EncryptedTypedSecret> retrieveAesSecrets(final String deviceIdentification,
+            final List<SecretType> secretTypes, SecretStatus status) {
         try {
-            //@formatter:off
-            return secretTypes.stream()
-                    .map(secretType -> this.retrieveSecret(deviceIdentification, secretType))
-                    .map(Optional::get)
-                    .collect(Collectors.toList());
-            //@formatter:on
+            return secretTypes.stream().map(secretType -> this.retrieveSecret(deviceIdentification, secretType, status))
+                              .collect(Collectors.toList());
         } catch (final Exception exc) {
             throw new IllegalStateException(
-                    String.format("Something went wrong retrieving secrets for device %s", deviceIdentification), exc);
+                    String.format("Something went wrong retrieving secrets for device %s: %s", deviceIdentification,
+                            exc.toString()), exc);
         }
     }
 
-    public Optional<TypedSecret> retrieveSecret(final String deviceIdentification, final SecretType secretType) {
-        final Date now = new Date();
-        final Long secretId = this.secretRepository.findIdOfValidMostRecent(deviceIdentification, secretType.name(),
-                now);
-        if (secretId == null) {
+    private EncryptedTypedSecret retrieveSecret(final String deviceIdentification, final SecretType secretType,
+            SecretStatus status) {
+        final Optional<DbEncryptedSecret> optional = this
+                .getSingleDbEncryptedSecret(deviceIdentification, secretType, status);
+        if (optional.isPresent()) {
+            return EncryptedTypedSecret.fromDbEncryptedSecret(optional.get());
+        } else {
+            return EncryptedTypedSecret.getNullInstance(secretType);
+        }
+    }
+
+    private Optional<DbEncryptedSecret> getSingleDbEncryptedSecret(final String deviceIdentification,
+            final SecretType secretType, final SecretStatus secretStatus) {
+        final List<DbEncryptedSecret> secretsList = this.secretRepository
+                .findSecrets(deviceIdentification, secretType, secretStatus);
+        boolean onlySingleSecretAllowed =
+                SecretStatus.NEW.equals(secretStatus) || SecretStatus.ACTIVE.equals(secretStatus);
+        if (secretsList.size() == 0) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1155)